### ISSUE TITLE
Pre-commit hook can't find node

### ISSUE THREAD

**Issue by bmv437** [ForkIssues/Tracker#311]
Hey there,

I'm using [husky](https://github.com/typicode/husky) to run a pre-commit hook on my repo, and everything runs correctly in my terminal when I run 
```git commit -m "some message"```

The issue is when I try and commit with Fork, and it attempts to run the pre-commit hook and fails because it can't find `node`. Here's the full error from Fork

```
Pre-commit hook 'pre-commit' has failed:
husky > npm run -s precommit (node )


husky > pre-commit hook failed (add --no-verify to bypass)

/Users/brandon/project/.git/hooks/pre-commit: line 49: node: command not found
env: node: No such file or directory
```

when I run `which node` I get
```/Users/brandon/.nvf/installed/node-8.5.0/bin/node```

Full disclosure I'm using macOS, with fish shell, and nvf to manage my node versions. I just really wonder how Fork is running the pre-commit in a way where it can't find my node installation. Let me know if you have any questions or need more info.

**Comment by DanPristupov** [comment]
It looks like Fork doesn't have node in PATH in ENV.
You can use absolute paths in your hook or you can run From from your terminal. In this way Fork will inherit PATH of the terminal and will be able to use same aliases.

**Comment by DanPristupov** [comment]
I suppose I can close this

**Comment by nicekiwi** [comment]
This is still an issue for me. When using Node Version manager, fork can't seem to find the npm executable.

**Comment by DanPristupov** [comment]
@nicekiwi did you try to start Fork from terminal as I mentioned in my comment?

Where is `npm` located in your system (`$ which npm`)?

**Comment by nicekiwi** [comment]
@DanPristupov yeah it does work starting form command line. 

`/Users/ezrasharp/.nvm/versions/node/v13.0.1/bin/npm`

**Comment by marianhlavac** [comment]
> You can use absolute paths in your hook or you can run From from your terminal. 

Well, I understand that the problem isn't trivial, but this isn't much elegant solution for the problem.

* Using absolute paths in hooks will break them for other developers in team.
* Running Fork always from terminal seems to be counter-intuitive, I like to have Fork on my macOS dock or run it using Spotlight.

I'd understand this behaviour when using bundled Fork git instance, but the system git instance should be equivalent to typing `git` into my default shell.

**Comment by DanPristupov** [comment]
> I'd understand this behaviour when using bundled Fork git instance, but the system git instance should be equivalent to typing git into my default shell.

No, things in OS don't work like that. In *nix (including BSD, including macOS) a process inherits environment of the parent process. The environment of your shell is available only for that shell process or the processes which it started. Fork (or any other process started from Dock or Spotlight) doesn't have an access to it.

**Comment by marianhlavac** [comment]
I understand how OS works and what process tree is, I was just stating what the end user would expect to happen when clicking commit in Fork, while using git hooks.

**Comment by McFlyssss** [comment]
I'm facing a similar issue. Fork is using a different node version than my project so then when running the tests with hooks they do not pass (They crash because of the node version).

Any suggestion guys?

**Comment by DanPristupov** [comment]
@McFlyssss Start Fork from the terminal using `open -a Fork` command.

**Comment by ospfranco** [comment]
First of all thanks @DanPristupov for the great app

In my case it fails to detect yarn, even though it is installed through volta, volta binaries are present on my $PATH variable:

```
/Users/osp/Developer osp@DeepThought >
echo $PATH               
/Users/osp/bin:/Users/osp/.volta/bin[...Other paths are here]
```

I agree with the previous comments, starting fork from the terminal is a non-solution, is there really no other way to make this work?

**Comment by Legys** [comment]
https://github.com/typicode/husky/issues/390#issuecomment-792619315 This worked fine for me, none of the existing solutions didn't help. They just change nothing.

**Comment by svvitale** [comment]
You can also symbolic link your npm installation to one of the directories listed in `/etc/paths` OR add the necessary profile paths to that file.  This avoids the need to start Fork from the terminal AND avoids modifying your pre-commit hooks.

**Comment by msgirlperl** [comment]
I always start fork from my terminal and it's still using the wrong version for me.  Any other ideas?  I tried `open -a fork` (I typically just type `fork`) but it still didn't work.

`which npm` yields:
`/Users/mcrowley/.nvm/versions/node/v12.14.1/bin/npm` for me.  

Attempting to commit through fork gives: 
```$ git commit --file=/var/folders/6v/d1928vt927n1_hgx7d53fxk00000gp/T/8D6E1136-B276-4FB6-93CC-1E5C0E3B6BBE
husky > pre-commit (node v13.7.0)
warning Integrity check: System parameters don't match
error Integrity check failed
error Found 1 errors.
husky > pre-commit hook failed (add --no-verify to bypass)
```

Full disclosure: I'm not super familiar with macs.

**Comment by DanPristupov** [comment]
@msgirlperl 
Try to add `echo $PATH` to the hook and run it from Fork. It will show what PATH is used.

**Comment by jogly** [comment]
> You can also symbolic link your npm installation to one of the directories listed in `/etc/paths` OR add the necessary profile paths to that file. This avoids the need to start Fork from the terminal AND avoids modifying your pre-commit hooks.

this worked for me on mac, for example:
```
ln -s $(which node) /usr/local/bin/node
```

**Comment by dgattey** [comment]
@jogly thanks, works perfectly and avoids having to mess with starting Fork some complicated way other than a dock icon

**Comment by certainlyakey** [comment]
This comment resolved the issue for me: https://github.com/typicode/husky/issues/390#issuecomment-1082345670

If you use a node version manager, basically, create `.huskyrc` in your home directory, insert a code there that inits NVM/fnm/etc in your Bash or ZSH profile, then set the default node version. 

For fnm it's:

```
eval "$(fnm env)"
fnm use default
```

**Comment by iethree** [comment]
I had a homebrew node installation conflicting with my fnm version node manager somehow. `brew uninstall node` cleared it all up for me.

**Comment by shelbyspeegle** [comment]
This ended up working for me (husky git hooks):

In the repo's `pre-commit` file, basically initialize nvm before using node / npx:
```
# Ignore errors because this may fail to load in some environments.
set +e; . $HOME/.nvm/nvm.sh; set -e
```

**Comment by ospfranco** [comment]
Still running into this in 2024. Besides my previous answer (symbolic link to /bin folder, which by the way, also fixes some XCode errors if you are using React Native or some other framework that uses node), the solutions by hook managers like Husky or LeftHook is an rc file.

Here is the documentation for [LeftHook RC File](https://github.com/evilmartians/lefthook/blob/master/docs/configuration.md#rc)

**Comment by ospfranco** [comment]
I found a generic way to fix this once and for all. Here is the tutorial:

https://ospfranco.com/make-mise-installed-tools-globally-available-for-gui-tools/

**Comment by LMFinney** [comment]
> This comment resolved the issue for me: [typicode/husky#390 (comment)](https://github.com/typicode/husky/issues/390#issuecomment-1082345670)
> 
> If you use a node version manager, basically, create `.huskyrc` in your home directory, insert a code there that inits NVM/fnm/etc in your Bash or ZSH profile, then set the default node version.
> 
> For fnm it's:
> 
> ```
> eval "$(fnm env)"
> fnm use default
> ```

Thanks for this.

FYI, I got an error with husky@9.1.7 that `.huskyrc` has been deprecated. Following the error message, I moved the file to `~/.config/husky/init.sh`, and it worked great.

**Comment by do-adams** [comment]
I solved it with https://github.com/fork-dev/Tracker/issues/996#issuecomment-2672425394 😃
